// Definition of protos and RPC services for Obiew.

syntax = "proto3";

package obiew;

message EmptyMessage {}

message ServerConfig {
    string my_addr = 1;
    string my_paxos = 2;
    repeated string replica = 3;
}

enum OperationType {
    CREATE = 0;
    UPDATE = 1;
    DELETE = 2;
}


// An Obiew Post.
message Post {
    int32 post_id = 1;
    int32 user_id = 2;
    string content = 3;
    int32 image_id = 4; // Id for image look-up.
    string created = 5;
    Post original_post = 6;
    int32 direct_repost_id = 7;
    int32 num_reposts = 8;
    int32 num_comments = 9;
    int32 num_likes = 10;
}

// Get Post(s) Request.
message GetPostsRequest {
    oneof id_oneof { // The request has to be one of the following types.
        int32 post_id = 1; // Get a post by post_id.
        int32 user_id = 2; // Get posts created by user_id.
        int32 feed_user_id =3; // Get posts in news feed for user_id.
    }
}

// Get Post(s) Response.
message GetPostsResponse {
    repeated Post post = 1;
}

// Set Post Request.
message SetPostRequest {
    Post post = 1;
    OperationType operation = 2;
}

// Set Post(s) Response.
message SetPostResponse {
    Post post = 1;
}

// UserStat displayed in User's Profile page.
message UserStat {
    int32 user_id = 1;
    int32 num_posts = 2;
    int32 num_following = 3;
    int32 num_followers = 4;
    repeated Post recent_posts = 5;
}

// Get UserStat Request.
message GetUserStatRequest {
    int32 user_id = 1;
}

// Get UserStat Response.
message GetUserStatResponse {
    UserStat user_stat = 1;
}

// A User.
message User {
    int32 user_id = 1;
    string password = 2; // TODO: Encryption.
    string name = 3;
    int32 profile_picture = 4; // Id for retrieving image.
    string email = 5;
    string phone = 6;
    int32 num_posts = 7;
    int32 num_following = 8;
    int32 num_followers = 9;
    repeated Post recent_posts = 10;
}

message RegisterRequest {
    User user = 1; // Require: Name, Password, ProfilePicture, Email, Phone
}

message RegisterResponse {
    User user = 1; // Contain: UserId, Name, ProfilePicture, Email, Phone
}

message LogInRequest {
    User user = 1; // Require: Name, Password
}

message LogInResponse {
    User user = 1; // Contain: UserId, Name, ProfilePicture, Email, Phone
}

// Get User Request.
message GetUserRequest {
    User user = 1; // Require: UserId
}

// Get User Response.
message GetUserResponse {
    User user = 1; // Contain: UserId, Name, ProfilePicture, Email, Phone
}

// Set User Request.
message SetUserRequest {
    User user = 1;
    OperationType operation = 2;
}

// Set User Response.
message SetUserResponse {
    User user = 1;
}

// Get Follow(s) Request.
message GetFollowsRequest {
    int32 follower_id = 1;
    int32 followee_id = 2;
}

// Get Follow(s) Response.
message GetFollowsResponse {
    repeated User user = 1;
}

// Set Follow Request.
message SetFollowRequest {
    int32 follower_id = 1;
    int32 followee_id = 2;
    OperationType operation = 3;
}

// A Repost under an Obiew Post.
message Repost {
    int32 repost_id = 1;
    int32 post_id = 2;
    int32 user_id = 3;
    string content = 4;
    string created = 5;
}

// Get Repost(s) Request.
message GetRepostsRequest {
    int32 post_id = 2; // Get reposts under post_id.
}

// Get Repost(s) Response.
message GetRepostsResponse {
    repeated Repost repost = 1;
}

// A Comment under an Obiew Post.
message Comment {
    int32 comment_id = 1;
    int32 post_id = 2;
    int32 user_id = 3;
    string content = 4;
    string created = 5;
}

// Get Comment(s) Request.
message GetCommentsRequest {
    oneof id_oneof { // The request has to be one of the following types.
        int32 comment_id = 1; // Get a comment by comment_id.
        int32 post_id = 2; // Get comments under post_id.
        int32 user_id = 3; // Get comments created by user_id.
    }
}

// Get Comment(s) Response.
message GetCommentsResponse {
    repeated Comment comment = 1;
}

// Set Comment Request.
message SetCommentRequest {
    Comment comment = 1;
    OperationType operation = 2;
}

message Like {
    int32 like_id = 1;
    int32 post_id = 2;
    int32 user_id = 3;
    string created = 4;
}

// Get Like(s) Request.
message GetLikesRequest {
    oneof id_oneof { // The request has to be one of the following types.
        int32 like_id = 1; // Get a like by like_id.
        int32 post_id = 2; // Get likes under post_id.
        int32 user_id = 3; // Get likes created by user_id.
    }
}

// Get Like(s) Response.
message GetLikesResponse {
    repeated Like like = 1;
}

// Set Like Request.
message SetLikeRequest {
    Like like = 1;
    OperationType operation = 2;
}

// RPC Service for posts, comments, likes, and follows.
service Obiew {

    rpc Register (RegisterRequest) returns (RegisterResponse) {}

    rpc LogIn (LogInRequest) returns (LogInResponse) {}

    rpc GetUser (GetUserRequest) returns (GetUserResponse) {}

    // rpc GetUserStat (GetUserStatRequest) returns (GetUserStatResponse) {}

    rpc GetFollows (GetFollowsRequest) returns (GetFollowsResponse) {}

    rpc GetPosts (GetPostsRequest) returns (GetPostsResponse) {}

    rpc GetReposts (GetRepostsRequest) returns (GetRepostsResponse) {}

    rpc GetComments (GetCommentsRequest) returns (GetCommentsResponse) {}

    rpc GetLikes (GetLikesRequest) returns (GetLikesResponse) {}

    rpc SetUser (SetUserRequest) returns (SetUserResponse) {}

    rpc SetFollow (SetFollowRequest) returns (EmptyMessage) {}

    rpc SetPost (SetPostRequest) returns (SetPostResponse) {}

    rpc SetComment (SetCommentRequest) returns (EmptyMessage) {}

    rpc SetLike (SetLikeRequest) returns (EmptyMessage) {}
}


service MultiPaxos {
    
    rpc Register (RegisterRequest) returns (RegisterResponse) {}

    rpc LogIn (LogInRequest) returns (LogInResponse) {}

    rpc GetUser (GetUserRequest) returns (GetUserResponse) {}

    rpc GetPosts (GetPostsRequest) returns (GetPostsResponse) {}

    rpc SetUser (SetUserRequest) returns (SetUserResponse) {}

    rpc SetPost (SetPostRequest) returns (SetPostResponse) {}
}